src/containerapp/azext_containerapp/containerapp_decorator.py

```diff
@@ -0,0 +1,713 @@
+    Volume as VolumeModel,
+    VolumeMount as VolumeMountModel)
+
+from ._utils import (_ensure_location_allowed,
+                     parse_secret_flags, store_as_secret_and_return_secret_ref, parse_env_var_flags,
+                     _convert_object_from_snake_to_camel_case,
+                     _object_to_dict, _remove_additional_attributes,
+                     _remove_readonly_attributes,
+                     _infer_acr_credentials,
+                     _ensure_identity_resource_id,
+                     validate_container_app_name, register_provider_if_needed,
+                     set_managed_identity,
+                     create_acrpull_role_assignment, is_registry_msi_system,
+                     safe_set, parse_metadata_flags, parse_auth_flags,
+                     get_default_workload_profile_name_from_env,
+                     ensure_workload_profile_supported, _generate_secret_volume_name,
+                     parse_service_bindings, check_unique_bindings, AppType,
+                     create_deserializer, process_loaded_yaml, load_yaml_file, get_linker_client,
+                     _validate_subscription_registered)
+from ._validators import validate_create, validate_revision_suffix
+
+from ._constants import (CONTAINER_APPS_RP,
+                         HELLO_WORLD_IMAGE)
+
+logger = get_logger(__name__)
+
+
+class BaseContainerAppDecorator:
+    def __init__(
+        self, cmd: AzCliCommand, client: Any, raw_parameters: Dict, models: str
+    ):
+        self.raw_param = raw_parameters
+        self.cmd = cmd
+        self.client = client
+        self.models = models
+
+    def register_provider(self):
+        register_provider_if_needed(self.cmd, CONTAINER_APPS_RP)
+
+    def validate_subscription_registered(self):
+        _validate_subscription_registered(self.cmd, CONTAINER_APPS_RP)
+
+    def validate_arguments(self):
+        raise NotImplementedError()
+
+    def construct_containerapp(self):
+        raise NotImplementedError()
+
+    def create_containerapp(self, containerapp_def):
+        raise NotImplementedError()
+
+    def construct_containerapp_for_post_process(self, containerapp_def, r):
+        raise NotImplementedError()
+
+    def post_process_containerapp(self, containerapp_def, r):
+        raise NotImplementedError()
+
+    def update_containerapp(self, containerapp_def):
+        raise NotImplementedError()
+
+    def list_containerapp(self):
+        try:
+            resource_group_name = self.get_argument_resource_group_name()
+            managed_env = self.get_argument_managed_env()
+            if self.get_argument_resource_group_name() is None:
+                containerapps = self.client.list_by_subscription(cmd=self.cmd)
+            else:
+                containerapps = self.client.list_by_resource_group(cmd=self.cmd, resource_group_name=resource_group_name)
+
+            if managed_env:
+                env_name = parse_resource_id(managed_env)["name"].lower()
```

Comment:
managed_env is always resource id?

Reply:
managed_env is always resource id.
in params setting : `validator=validate_managed_env_name_or_id`
it will help to construct managed_env to a resource id.